src/meta/Intersection.ts

```diff
@@ -0,0 +1,89 @@
+import global from '@dojo/shim/global';
+import WeakMap from '@dojo/shim/WeakMap';
+import Map from '@dojo/shim/Map';
+import { createHandle } from '@dojo/core/lang';
+import { Base } from './Base';
+
+import 'intersection-observer';
+
+interface ExtendedIntersectionObserverEntry extends IntersectionObserverEntry {
+	readonly isIntersecting: boolean;
+}
+
+interface IntersectionDetail extends IntersectionGetOptions {
+	entries: WeakMap<Element, IntersectionResult>;
+	observer: IntersectionObserver;
+}
+
+export interface IntersectionGetOptions {
+	root?: string;
+	rootMargin?: string;
+	threshold?: number[];
+}
+
+export interface IntersectionResult {
+	intersectionRatio: number;
+	isIntersecting: boolean;
+}
+
+const defaultIntersection: IntersectionResult = Object.freeze({
+	intersectionRatio: 0,
+	isIntersecting: false
+});
+
+export class Intersection extends Base {
+	private readonly _details: Map<string, IntersectionDetail> = new Map();
+
+	public get(key: string, options: IntersectionGetOptions = {}): IntersectionResult {
+		let rootNode: HTMLElement | undefined;
+		if (options.root) {
+			rootNode = this.getNode(options.root);
+			if (!rootNode) {
+				return defaultIntersection;
+			}
+		}
+		const node = this.getNode(key);
+		if (!node) {
+			return defaultIntersection;
+		}
+
+		let details = this._getDetails(options);
+		if (!details) {
+			details = this._createDetails(options, rootNode);
+			details.observer.observe(node);
+		}
+
+		return details.entries.get(node) || defaultIntersection;
+	}
+
+	public has(key: string | number, options?: IntersectionGetOptions): boolean {
+		const node = this.getNode(key);
+		const details = this._getDetails(options);
+		return Boolean(details && node && details.entries.has(node));
+	}
+
+	private _createDetails(options: IntersectionGetOptions, rootNode?: HTMLElement): IntersectionDetail {
+		const entries = new WeakMap<HTMLElement, ExtendedIntersectionObserverEntry>();
+		const observer = new global.IntersectionObserver(this._onIntersect(entries), { ...options, root: rootNode });
+		const details = { observer, entries, ...options };
+
+		this._details.set(JSON.stringify(options), details);
+		this.own(createHandle(observer.disconnect));
+		return details;
+	}
+
+	private _getDetails(options: IntersectionGetOptions = {}): IntersectionDetail | undefined {
+		return this._details.get(JSON.stringify(options));
```

Comment:
I am a bit concerned about GC with this... I can understand why stringifying is useful to create a _key_ here, but it is a map, which means it will have strong references to the values, which means that they won't ever get released.  Shouldn't here be a destruction handle in the constructor to `.clear()` the Map?

Reply:
We use `Map`s across the widget-core, should we raise an separate issue to discuss this?